main/modes/factoryTest/factoryTest.c

```diff
@@ -0,0 +1,804 @@
+        case 4:
+            return TB_DOWN;
+        case 5:
+            return TB_UP_RIGHT;
+        case 6:
+            return TB_UP_LEFT;
+        case 7:
+            return TB_DOWN_LEFT;
+        case 8:
+            return TB_DOWN_RIGHT;
+        default:
+            return UINT8_MAX;
+    }
+}
+
+/**
+ * @brief Read, save, and validate touch
+ *
+ */
+void testReadAndValidateTouch(void)
+{
+    int32_t phi, r, intensity;
+    test->touched = getTouchJoystick(&phi, &r, &intensity);
+
+    // Transition states
+    if (test->touched)
+    {
+        test->touchJoystick = getTouchJoystickZones(phi, r, true, true);
+        uint8_t pad         = touchJoystickToTouchIdx(test->touchJoystick);
+        if (pad == UINT8_MAX)
+        {
+            return;
+        }
+
+        if (BTN_NOT_PRESSED == test->touchStates[pad])
+        {
+            test->touchStates[pad] = BTN_PRESSED;
+        }
+
+        if (test->lastTouchStateIdx != UINT8_MAX && pad != test->lastTouchStateIdx)
+        {
+            if (BTN_PRESSED == test->touchStates[test->lastTouchStateIdx])
+            {
+                test->touchStates[test->lastTouchStateIdx] = BTN_RELEASED;
+            }
+        }
+
+        test->lastTouchStateIdx = pad;
+    }
+    else if (test->lastTouchStateIdx != UINT8_MAX)
+    {
+        if (BTN_PRESSED == test->touchStates[test->lastTouchStateIdx])
+        {
+            test->touchStates[test->lastTouchStateIdx] = BTN_RELEASED;
+        }
+    }
+
+    // Check if all buttons have passed
+    test->touchPassed = true;
+    for (uint8_t i = 0; i < 9; i++)
+    {
+        if (BTN_RELEASED != test->touchStates[i])
+        {
+            test->touchPassed = false;
+        }
+    }
+}
+
+/**
+ * @brief Audio callback. Take the samples and pass them to test
+ *
+ * @param samples The samples to process
+ * @param sampleCnt The number of samples to process
+ */
+void testAudioCb(uint16_t* samples, uint32_t sampleCnt)
+{
+    // For each sample
+    for (uint32_t idx = 0; idx < sampleCnt; idx++)
+    {
+        // Push to test
+        PushSample32(&test->dd, samples[idx]);
+
+        // If 128 samples have been pushed
+        test->samplesProcessed++;
+        if (test->samplesProcessed >= 128)
+        {
+            // Update LEDs
+            test->samplesProcessed = 0;
+            HandleFrameInfo(&test->end, &test->dd);
+
+            // Keep track of max value for the spectrogram
+            int16_t maxVal = 0;
+            for (uint16_t i = 0; i < FIX_BINS; i++)
+            {
+                if (test->end.fuzzed_bins[i] > maxVal)
+                {
+                    maxVal = test->end.fuzzed_bins[i];
+                }
+            }
+
+            // If already passed, just return
+            if (true == test->bzrMicPassed)
+            {
+                return;
+            }
+        }
+    }
+}
+
+/**
+ * @brief Read, save, and validate accelerometer data
+ *
+ */
+void testReadAndValidateAccelerometer(void)
```

Comment:
@cnlohr after this gets merged to main and you want to add your motion and calibration stuff, this is likely the right place to do it